### ISSUE TITLE
Update the volume blocklet automatically

### ISSUE THREAD

**Issue by enanajmain** [vivien/i3blocks#351]
I have a volume blocklet with the community contributed [volume module](https://github.com/vivien/i3blocks-contrib/blob/master/volume/volume) with the following i3blocks config:

    [volume]
    label=
    instance=Master
    command=~/.config/i3/scripts/volume
    interval=once
    signal=10
    border=#e06c75

Now, I've set the following keybindings in i3 config file:

    bindsym XF86AudioRaiseVolume exec --no-startup-id amixer -q set Master 1%+ && pkill -RTMIN+10 i3blocks
    bindsym XF86AudioLowerVolume exec --no-startup-id amixer -q set Master 1%- && pkill -RTMIN+10 i3blocks
    bindsym XF86AudioMute exec --no-startup-id amixer -q -D pulse set Master toggle && pkill -RTMIN+10 i3blocks

So, when I change the volume by pressing these keys the i3bar updates itself. But in my laptop, I often plug in and out my headphone, which causes the volume to change. But since the blocklet is set to update when receiving a signal 10, the i3bar doesn't update in this case. Each time I plugin in/out my headphone I have to press any of the bound keys to update the i3bar. I had used i3status before, and I saw that it updates when I plug in/out my headphone, and that's instantaneous, even though I have set the update time of i3status to 5 seconds. Is there any way the same can be done with i3blocks?

**Comment by gportay** [comment]
I am running `amixer -q set Master 1%+ && pkill -RTMIN+10 i3blocks` in a terminal and it works fine on my laptop. Did I miss something?

Which version of i3blocks are you using?

Can you update or mute the volume in the bar using the mouse?

**Comment by enanajmain** [comment]
@gportay, it works if you use the mouse buttons or keyboard key to change the volume. Because I've explicitly bound them to do so. But the volume also changes, without pressing any buttons in either mouse or keyboard, when I plug in or plug out my headphone. Because it changes the output device from speaker to headphone. But in this particular case, my i3blocks doesn't update.

**Comment by kb100** [comment]
For future reference, @enanajmain, issues with blocks in i3blocks-contrib should be posted as issues to i3blocks-contrib. i3blocks issues are for problems with the scheduler itself, not with blocks.

In any case, I'm happy to help you here. Could you clarify one thing for me please? Does the block pick up your headphone after the next time it does update, say by you manually hitting the volume up button on your keyboard? Or does it still say your previous audio device even after an update?

Your issue could potentially be fixed by using the volume-pulseaudio blocklet with SUBSCRIBE=1.

**Comment by enanajmain** [comment]
@kb100, I'll keep i3blocks-contrib in mind. And about the query, the block does *not* pick up my headphone after the next time it updates.

To lay out the whole scenario, say I power up my laptop. The volume is 85% in speaker mode. Then I plug in my headphone. The block still says 85% speaker mode even though the device has been switched to headphone automatically. I have checked with alsamixer and also just by playing audio. Then I hit mute button. The block updates to show that the device is muted. I again press mute button, the block updates to show 100% in headphone mode. I plug out the headphone. The block still shows 100% volume. I press mute button and the block updates and shows muted speaker mode. I press mute button again and the block shows 85% in speaker. I again plug in my headphone. The block still does *not* update.

To summarize:
power up                 > actually 85% speaker + shows 85% speaker
headphone plug in  > actually 100% headphone + shows 85% speaker
press mute button   > actually muted headphone + shows muted headphone
press mute button   > actually 100% headphone + shows 100% headphone
plug out headphone> actually 100% speaker + shows muted headphone
press mute button   > actually muted speaker + shows muted speaker
press mute button   > actually 85% speaker + shows 85% speaker
plug in headphone  > actually 100% headphone + shows 85% speaker i.e. still not updating

To remind you, I checked i3status, and it updates volume block with headphone plugging in/out automatically, if that helps somehow.

**Comment by kb100** [comment]
Thanks for the extra info, this helps a lot. There is one inconsistency that doesn't add up yet. You say "the block does not pick up my headphone after the next time it updates". However, as far as I can tell, nothing in your description or summary shows something incorrect AFTER a block update. For completeness, let me go through your summary.

power up > actually 85% speaker + shows 85% speaker (CORRECT)
headphone plug in > actually 100% headphone + shows 85% speaker (INCORRECT, but before update)
press mute button > actually muted headphone + shows muted headphone (CORRECT, pressing mute caused block update)
press mute button > actually 100% headphone + shows 100% headphone (CORRECT)
plug out headphone> actually 100% speaker + shows muted headphone (INCORRECT, but before update)
press mute button > actually muted speaker + shows muted speaker (CORRECT, pressing mute caused block update)
press mute button > actually 85% speaker + shows 85% speaker (CORRECT)
plug in headphone > actually 100% headphone + shows 85% speaker i.e. still not updating (INCORRECT, but before update)

Note that not all blocks update simultaneously, so just because your bar changed, that doesn't mean the volume block updated. That said, obviously something is wrong. The reason I have to determine whether the wrongness happens before or after a block update is that if it happens AFTER, then there is an error in how the block is processing your system volume and or device, but if it happens BEFORE, it is most likely an issue with the user's configuration not matching their expectation. The latter is much easier to solve with improved documentation and instruction.

If there is something I have missed, please let me know. Otherwise, I think this is an expectation mismatch problem. The basic volume blocklet does not subscribe to volume events, which is (if I recall correctly) how i3status automatically updates. Instead it only waits for user signal to update. If you wanted to make a minimal change to your configuration, I believe that changing the config to

````
[volume]
label=
instance=Master
command=~/.config/i3/scripts/volume
interval=5
signal=10
border=#e06c75
````

would solve your issue, assuming that second resolution in your volume is acceptable to you. The other option I would recommend is to use the more sophisticated volume-pulseaudio blocklet with the SUBSCRIBE=1 option. The subscribe option uses a callback mechanism to have pulseaudio notify the blocklet whenever there is a volume update, device change, etc. This option would allow you to get immediate updates to the block as soon as any audio device changes.

**Comment by gportay** [comment]
@enanajmain: okay, this is normal.

The blocklet cannot be notified that someone else has changed the volume, unless you use a signal to prevent i3blocks to respawn your blocket that triggers an update of its output string.

**Comment by gportay** [comment]
I tried to set `interval` to 1 second so i3blocks respawn the volume blocklet, but it did not worked :/

**Comment by gportay** [comment]
@kb100: I had a quick look to i3status and it does not seem to register to volume change. I assume it polls every second :/

**Comment by enanajmain** [comment]
@kb100, first of all, the volume-pulseaudio works, although that is a little slow.

And about your query, the block is correct after updating. you are right. But my problem was that I had to manually update the block, you see. So, if I somehow forgot to do so, the block will go on having an incorrect info because I didn't manually updated the block by pressing mute/volume up/volume down button. That's because I set interval to once, and then the block would only update when I press those bound keys in i3 config file. There is no problem with i3config or the block itself. I just thought it should also automatically check for audio setting changes, just like volume-pulseaudio does.

Thanks for the help.

**Comment by vivien** [comment]
Glad to see that *volume-pulseaudio* does work for you.
FWIW, I prefer to split the command and signal on key release, like:

````
bindsym XF86AudioLowerVolume exec --no-startup-id amixer -q set Master 1%-
bindsym XF86AudioLowerVolume --release pkill -RTMIN+10 i3blocks
````

**Comment by enanajmain** [comment]
@vivien, I had no idea you could do that. Thanks. Where do I get these options? I learnt about exec from a youtube video about i3wm.
